src/github.com/getlantern/detour/ioloop.go

```diff
@@ -0,0 +1,315 @@
+package detour
+
+import (
+	"bytes"
+	"errors"
+	"sync/atomic"
+)
+
+// to pass result of read operation to merger from underlie connection
+type innerReadResult struct {
+	// the connection which performs this read
+	c conn
+	// buffer holding received data, it's not necessarily the same
+	// buffer as innerReadRequest.
+	buf []byte
+	// IO error, if any
+	err error
+}
+
+type innerReadRequest struct {
+	buf      []byte
+	chResult chan innerReadResult
+}
+
+// ioLoop is the core of detour. It waits for connections and handles
+// read/write requests
+func (dc *Conn) ioLoop() {
+	// Use buffered channel in same goroutine so we can easily add / remove
+	// connections. Should switch to container/ring if performace matters.
+	chConns := make(chan conn, 2)
+	// Requests ioLoop to remove already closed connections.
+	chRemoveConn := make(chan conn)
+	// Hold the read request so we can re-read after replay.
+	var firstReadReq *innerReadRequest
+
+	// Can we replay this request
+	// TODO: How to check when using pipelining or HTTP/2?
+	var nonidempotentHTTPRequest bool
+
+	for {
+		select {
+		case c := <-dc.chConnToIOLoop:
+			if dc.anyDataReceived() {
+				log.Tracef("%s connection to %s available after data received, closing", c.Type(), dc.addr)
+				closeConn(c)
+				continue
+			}
+			reRead := false
+			if firstReadReq != nil {
+				if nonidempotentHTTPRequest {
+					log.Tracef("Not replay nonidempotent request to %s, only add to whitelist", dc.addr)
+					AddToWl(dc.addr, false)
+					closeConn(c)
+					continue
+				}
+				if !dc.replay(c) {
+					closeConn(c)
+					continue
+				}
+				reRead = true
+			}
+			chConns <- c
+			// spawn goroutine after above statement so channel
+			// operations are in determined order.
+			if reRead {
+				go func() {
+					log.Tracef("Re-read from %s connection to %s", c.Type(), dc.addr)
+					buf := make([]byte, len(firstReadReq.buf))
```

Comment:
We need separate buffers to avoid race condition. `sync.Pool` can be employed to avoid frequent memory allocation, but let's leave it for future optimization.